src/py21cmfast/src/HaloBox.c

```diff
@@ -1,105 +1,276 @@
-    double f10 = astro_params_stoc->F_STAR10;
-    double fa = astro_params_stoc->ALPHA_STAR;
-    double sigma_star = astro_params_stoc->SIGMA_STAR;
-    double sigma_sfr = astro_params_stoc->SIGMA_SFR;
-
-    double f7 = astro_params_stoc->F_STAR7_MINI;
-    double fa_m = astro_params_stoc->ALPHA_STAR_MINI;
-    double fesc7 = astro_params_stoc->F_ESC7_MINI;
-
-    double fstar_mean, fstar_mean_mini, sfr_mean, sfr_mean_mini;
-    double f_sample, f_sample_mini, sm_sample, n_ion_sample, sfr_sample, wsfr_sample;
-    double f_rng, sfr_rng;
-    double sm_sample_mini, sfr_sample_mini;
-    double fesc,fesc_mini;
+    if(flag_options->FIXED_HALO_GRIDS || user_params_stoc->AVG_BELOW_SAMPLER){
+        if(flag_options->PHOTON_CONS_TYPE == 2)
+            consts->alpha_esc = get_fesc_fit(redshift);
+        else if(flag_options->PHOTON_CONS_TYPE == 3)
+            consts->fesc_10 = get_fesc_fit(redshift);
 
-    fesc = fmin(norm_esc_var*pow(halo_mass/1e10,alpha_esc_var),1);
+        consts->Mlim_Fstar = Mass_limit_bisection(global_params.M_MIN_INTEGRAL, global_params.M_MAX_INTEGRAL, consts->alpha_star, consts->fstar_10);
+        consts->Mlim_Fesc = Mass_limit_bisection(global_params.M_MIN_INTEGRAL, global_params.M_MAX_INTEGRAL, consts->alpha_esc, consts->fesc_10);
+
+        if(flag_options->USE_MINI_HALOS){
+            consts->Mlim_Fstar_mini = Mass_limit_bisection(global_params.M_MIN_INTEGRAL, global_params.M_MAX_INTEGRAL, consts->alpha_star_mini,
+                                                            consts->fstar_7 * pow(1e3,consts->alpha_star_mini));
+            consts->Mlim_Fesc_mini = Mass_limit_bisection(global_params.M_MIN_INTEGRAL, global_params.M_MAX_INTEGRAL, consts->alpha_esc,
+                                                            consts->fesc_7 * pow(1e3,consts->alpha_esc));
+        }
+    }
+}
+
+void get_halo_stellarmass(double halo_mass, double mturn_acg, double mturn_mcg, double star_rng,
+                             struct HaloBoxConstants *consts, double *star_acg, double *star_mcg){
+    //low-mass ACG power-law parameters
+    double f_10 = consts->fstar_10;//TODO: pass in log
+    double f_a = consts->alpha_star;
+    double sigma_star = consts->sigma_star;
+
+    //high-mass ACG power-law parameters
+    double fu_a = consts->alpha_upper;
+    double fu_p = consts->pivot_upper;//TODO: pass in log
+
+    //MCG parameters
+    double f_7 = consts->fstar_7;
+    double f_a_mini = consts->alpha_star_mini;
+
+    //intermediates
+    double fstar_mean;
+    double f_sample, f_sample_mini;
+    double sm_sample,sm_sample_mini;
+
+    double baryon_ratio = cosmo_params_stoc->OMb / cosmo_params_stoc->OMm;
+
+    /* Simply adding lognormal scatter to a delta increases the mean (2* is as likely as 0.5*)
+    * We multiply by exp(-sigma^2/2) so that X = exp(mu + N(0,1)*sigma) has the desired mean */
+    star_rng = -sigma_star*sigma_star/2 + star_rng*sigma_star;
```

Comment:
Hold on -- is `star_rng` a random number _generator_ here? Or is it a random number drawn from N(0,1)? Also, I'm not sure I quite understand the notation -- is `-sigma^2/2` equal to your `mu`? Also, is adjusting the distribution so its mean in linear space is correct the right thing to do? Are we sure that the given mean doesn't represent the log-space mean?

Reply:
This is the adjustment from the mean of the halo property to the mean log of the halo property. I've set the parameters up to be the mean of whatever distribution you want, so that the mean doesn't change when adjusting the sigmas.

as for the `_rng` notation, you're right this is a little confusing, these are the correlated random numbers drawn from the normal distribution, so that when we multiply `(mean * exp(X_rng))` we get the lognormal scatter.

This function is not super clear, since the steps were all shuffled around to minimise exp/log calls.

Reply:
One can argue whether the parameter should represent the mean in linear or log. I'm open to suggestion either way but zeroth order I would think that someone just changing the sigma shouldn't expect the global history to shift dramatically.

Reply:
Yes, I guess that is the correct criterion -- that the change in observables should be small given a change in the sigma(s). Thanks for adding some docs to `inputs.py` to clarify it.

Reply:
After talking with Andrei I've gone with your option (no -sigma^2, F_STAR10 is the mean of the log), the main reason being that all the observational scaling relation fits are done in log-log space anyway.